### PR TITLE
Implemented major datatypes used in the world

### LINKED ISSUE #41: Implement World datastructures
Now that interfaces are finalized, we need to implement the data structures so we can actually test them and use them when implementing other new features.

This includes simple implementations of the following:
* Robot
* Ball
* Team
* World

### PR DESCRIPTION
**Pull request by MathewMacDougall**
Added implementations for the `Ball`, `Field`, `Team`, and `World` so we can properly use them to implement and test other stuff. Also added unit tests.

resolves #41

### CODE REVIEW

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Can we break this up? In general anything over a few hundred lines is difficult for me to review effectively, and there's research to support anything longer then a few hundred lines is dramatically less effective, both for the reviewer and author:

https://smallbusinessprogramming.com/optimal-pull-request-size/
https://robertheaton.com/2015/10/26/why-and-how-to-make-smaller-pull-requests/
https://opensource.com/sites/default/files/uploads/pull_request_size_view_time_0.png

File: src/thunderbots/software/ai/world/team.cpp [hunk 1]
```diff
@@ -1,33 +1,134 @@
...
-Team::Team()
+// We initialize the goalie_id to -1, a value that will never occur naturally as all robot
+// ids are numbers >= 0 because otherwise the goalie_id will be initialized to 0 by
+// default, and we want there to be no goalie until the value is updated.
+Team::Team() : team_robots(), goalie_id(-1)
 {
 }
 
-void Team::update(std::vector<Robot> &team_robots)
+void Team::update(const thunderbots_msgs::Team& team_msg)
```
Comment (jordanzliu) [inline src/thunderbots/software/ai/world/team.cpp:12]: imo we could eliminate our code's coupling with ROS here and keep everything that touches ROS in main.cpp.

Comment (MathewMacDougall) [reply src/thunderbots/software/ai/world/team.cpp:12]: Will look into it. Going to close this PR and break it up.

**Comment by MathewMacDougall** [comment]
Sorry didn't realize my line count got that big. Will break it up.

**Pull request closed without merging**